nav.1
.TH NAV 1
.SH NAME
nav \- a highly configurable, performant, async system tool
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBnav\fR [\fB\-\-version\fR] [\fB\-\-help\fR] [\fB\-\-debug\fR]
.TP
\&\fBnav\fR [\fB\-\-config\fR=\fIpath\fR] [\fB\-\-log-file\fR=\fIpath\fR] [\fB\-\-verbose\fR=\fIgroup\fR]
.SH DESCRIPTION
nav is a hackable ncurses file manager, inspired by vim, ranger, and dvtm.

nav searches for a config file in the following order:
  \fB~/.navrc\fP
  \fB~/.nav/.navrc\fP
  \fB/etc/navrc\fP

.SH OPTIONS
.IP "\fB\-c\fR \fIpath\fR, \fB\-\-config\fR=\fIpath\fR" 14
.IX Item "--config=path"
Specify an alternative config file.
.IP "\fB\-l\fR \fIpath\fR, \fB\-\-log-file\fR=\fIpath\fR" 14
.IX Item "-l, --log-file"
Specify an alternative log file
(Sets \fB--debug\fP).
.IP "\fB\-V\fR \fIgroup\fR, \fB\-\-verbose\fR=\fIgroup\fR" 14
Verbose mode for a specific log group
(Sets \fB--debug\fP).
.IX Item "-V, --verbose"
.IP "\fB\-d\fR, \fB\-\-debug\fR" 14
.IX Item "--debug"
Debug mode.
.IP "\fB\-v\fR, \fB\-\-version\fR" 14
.IX Item "--version"
Print version information and exit.
.IP "\fB\-h\fR, \fB\-\-help\fR" 14
.IX Item "-h, --help"
Print this help message and exit.

.SH DEFAULT KEYS
.SS "\s-1Buffer\s0"
.IP "\fBh\fR"
go left
.IP "\fBj\fR"
go down
.IP "\fBk\fR"
go up
.IP "\fBl\fR"
go right
.IP "\fBC-j\fR"
jump page down
.IP "\fBC-k\fR"
jump page up
.IP "\fBn\fR"
next search result
.IP "\fBN\fR"
previous search result
.IP "\fBC-o\fR"
jumplist cycle backward
.IP "\fBC-i\fR"
jumplist cycle forward
.IP "\fBf\fR"
toggle filter
.IP "\fBg\fR"
g operator
.IP "\fBG\fR"
goto bottom
.IP "\fBV\fR"
toggle visual select
.IP "\fBy\fR"
yank operator
.IP "\fBd\fR"
cut operator
.IP "\fBm\fR"
mark operator
.IP "\fB’\fR"
jump operator
.IP "\fBp\fR"
paste
.IP "\fBX\fR"
remove

.RE
.SS "\s-1Window\s0"
.IP "\fBH\fR"
move window focus left
.IP "\fBJ\fR"
move window focus down
.IP "\fBK\fR"
move window focus up
.IP "\fBL\fR"
move window focus down
.IP "\fB:\fR"
start command-line mode
.IP "\fB/\fR"
forward search
.IP "\fB?\fR"
backward search

.RE
.SS "\s-1Term\s0"
.IP "\fBMeta-[\fR"
escape terminal focus
.IP "\fBMeta-k\fR"
scroll buffer up
.IP "\fBMeta-j\fR"
scroll buffer down

.RE
.SS "\s-1Command-line\s0"
.IP "\fBESC\fR"
cancel/close ex_mode
.IP "\fBTAB\fR"
compl cycle forward
.IP "\fBS-TAB\fR"
compl cycle backward
.IP "\fBCAR\fR"
run cmd
.IP "\fBBS\fR"
(internal)
.IP "\fBSpc\fR"
(internal)
.IP "\fBC-p\fR"
history cycle backward
.IP "\fBC-n\fR"
history cycle forward
.IP "\fBC-w\fR"
delete word
.IP "\fBC-b\fR"
move cursor back word
.IP "\fBC-e\fR"
move cursor forward word
.IP "\fBC-s\fR"
move cursor start of line
.IP "\fBC-f\fR"
move cursor end of line
.IP "\fBC-u\fR"
delete line
.IP "\fBC-l\fR"
add invert symbol to cmd
.IP "\fBC-g\fR"
hint mode
.IP "\fBC-j\fR"
go down in menu
.IP "\fBC-k\fR"
go up in menu

.RE
.SS "\s-1Operator\s0"
.IP "\fBm\fR {\fIa-zA-Z0-9\fR}"
set mark {\fIa-zA-Z0-9\fR} for path.
.IP "\fB’\fR {\fIa-zA-Z0-9\fR}"
jump to the mark {\fIa-zA-Z0-9\fR}
.IP "\fBC-w\fR {\fBHJKL\fR}"
swap buffer in direction
.IP "\fByy\fR
yank visible field
.IP "\fByf\fR
yank fullpath
.IP "\fByn\fR
yank name
.IP "\fByN\fR
yank name (no extension)
.IP "\fBye\fR"
yank extension
.IP "\fByd\fR"
yank directory
.IP "\fByt\fR"
yank type (opgroup)
.IP "\fByk\fR"
yank kind (file type)
.IP "\fBdd\fR"
yank to delete register

.RE
.SS "\s-1Symbol Expansion\s0"
.IP "\fB%b\fR"
buffer number
.IP "\fB%B\fR"
buffer name
.IP "\fB%f\fR"
fullpath
.IP "\fB%n\fR"
name
.IP "\fB%N\fR"
name (no extension)
.IP "\fB%e\fR"
extension
.IP "\fB%d\fR"
directory
.IP "\fB%t\fR"
type (opgroup)
.IP "\fB%k\fR"
kind (file type)
.IP "\fB%:%\fR"
nav pid
.IP "\fB%:!\fR"
the pid of the last !cmd
.IP "\fB%:?\fR"
the exit status of the last !cmd
.IP "\fB%_:<field>\fR"
selection field value
.IP "\fB%o:<var>\fR"
opgroup variable

.SH FILES
User-local info file: \fI~/.navinfo\fR.
.TP
Global configuration file: \fI/etc/navrc\fR.
.SH BUGS
If you find a bug, please report it at
.br
<\fBhttp://github.com/jollywho/nav\fP>.
.SH AUTHORS
Kevin Vollmer <kvollmer@metaphorize.net>


src/nav/nav.c
#include <string.h>
#include <curses.h>
#include <locale.h>
#include <unistd.h>
#include <getopt.h>

#include "nav/log.h"
#include "nav/nav.h"
#include "nav/option.h"
#include "nav/config.h"
#include "nav/tui/screen.h"
#include "nav/tui/window.h"
#include "nav/tui/ex_cmd.h"
#include "nav/event/event.h"
#include "nav/event/input.h"
#include "nav/tui/history.h"
#include "nav/table.h"
#include "nav/compl.h"
#include "nav/event/hook.h"
#include "nav/event/shell.h"
#include "nav/event/file.h"
#include "nav/vt/vt.h"

void init(int debug, char *config_path)
{
  log_msg("INIT", "INIT_START");
  if (debug)
    log_init();

  setlocale(LC_CTYPE, "");
  char *term = getenv("TERM");

  log_msg("INIT", "initscr");
  initscr();
  start_color();
  use_default_colors();
  noecho();
  nonl();
  curs_set(0);
  //raw();
  vt_init(term);

  cmd_init();
  option_init();
  tables_init();
  event_init();
  input_init();
  compl_init();
  hook_init();
  ex_cmd_init();
  window_init();
  file_init();

  config_init();
  hist_init();
  config_load_info();
  config_load(config_path);

  screen_init();
  cmd_sort_cmds();
  log_msg("INIT", "INIT_END");
}

void cleanup(void)
{
  log_msg("CLEANUP", "CLEANUP_START");
  ex_cmd_cleanup();
  hist_cleanup();
  file_cleanup();
  window_cleanup();
  hook_cleanup();
  compl_cleanup();
  input_cleanup();
  event_cleanup();
  tables_cleanup();
  option_cleanup();
  cmd_cleanup();
  vt_shutdown();
  endwin();
  log_msg("CLEANUP", "CLEANUP_END");
  log_cleanup();
}

static const char* usage =
"Usage: nav [options] [command]\n"
"\n"
"  -c, --config=<config>  Specify an alternative config file.\n"
"  -l, --log-file=<path>  Specify an alternative log file (Requires -d).\n"
"  -V, --verbose=<group>  Verbose mode for a specific log group.\n"
"  -d, --debug            Debug mode.\n"
"  -v, --version          Print version information and exit.\n"
"  -h, --help             Print this help message and exit.\n"
"\n";

static struct option long_options[] = {
  {"config", required_argument, NULL, 'c'},
  {"log-file", required_argument, NULL, 'l'},
  {"verbose", required_argument, NULL, 'V'},
  {"debug", no_argument, NULL, 'd'},
  {"version", no_argument, NULL, 'v'},
  {"help", no_argument, NULL, 'h'},
};

int main(int argc, char **argv)
{
  int debug = 0;
  char *config_path = NULL;

  for (int i = 0; i < argc; i++) {
    int option_index = 0;
    int c = getopt_long(argc, argv, "c:l:V:dvh", long_options, &option_index);
    if (c < 0)
      continue;

    switch (c) {
      case 'h':
        fprintf(stdout, "%s", usage);
        exit(EXIT_SUCCESS);
        break;
      case 'c':
        config_path = optarg;
        break;
      case 'l':
        debug = 1;
        log_set_file(optarg);
        break;
      case 'd':
        debug = 1;
        break;
      case 'V':
        debug = 1;
        log_set_group(optarg);
        break;
      case 'v':
        fprintf(stdout, "%s\n", NAV_LONG_VERSION);
        exit(EXIT_SUCCESS);
        break;
      default:
        fprintf(stderr, "%s", usage);
        exit(EXIT_FAILURE);
    }
  }

  init(debug, config_path);
  start_event_loop();
  DO_EVENTS_UNTIL(!mainloop_busy());
  config_write_info();
  cleanup();
}
